src/helpers/buffers/fsv.rs

```diff
@@ -0,0 +1,220 @@
+use bitvec::bitvec;
+use bitvec::prelude::BitVec;
+use std::ops::Range;
+
+/// A vector of bytes that never grows over a certain size or shrinks below that size.
+/// Vector is segmented into regions and `max_size` is a factor of number of regions.
+/// For example, if number of regions is `2`, then `max_size` must be a multiple of two.
+///
+/// Elements can be added to this vector in random order as long as the total number of elements
+/// does not exceed the overall capacity of the vector. An attempt to do so, will lead to a panic.
+/// Every element is stored by a specific offset and it is assumed that all elements are of the same
+/// size.
+///
+/// The layout of the vector with some elements added to it is presented below. In this example,
+/// vector has 3 regions and `X` indicates that space at that element is occupied.
+///
+///  region1  region2  region3
+/// [X,_,_,_][_,_,X,_][_,X,_,X]
+///
+/// Once `region1` is completely filled up, it is possible to drain the vector
+///
+/// [X,X,X,X][_,_,X,_][_,X,_,X] -> `drain` -> [_,_,X,_][_,X,_,X][_,_,_,_]
+///
+/// This vector is used inside the send buffer to keep track of messages added to it. Once first
+/// batch of messages is ready (region1 is full), it drains this vector and send those messages
+/// down to the network layer
+#[derive(Debug)]
+pub struct FixedSizeByteVec<const N: usize> {
+    data: Vec<u8>,
+    region_size: usize,
+    added: BitVec,
+    elements_drained: usize,
+}
+
+impl<const N: usize> FixedSizeByteVec<N> {
+    pub const ELEMENT_SIZE_BYTES: usize = N;
+
+    pub fn new(region_count: usize, region_size: usize) -> Self {
+        Self {
+            data: vec![0_u8; N * region_size * region_count],
+            added: bitvec![0; region_size * region_count],
+            elements_drained: 0,
+            region_size,
```

Comment:
It might be neater in some ways if you allow the caller to take an arbitrary number of elements; then the drain/take could accept an argument that says how many records you want to return.  Nothing in this design requires that this object remember how many records are returned in each pull.  That means that - if you want to force progress, you can take a small slice - and on every pull you might pull more if more is available.

Reply:
true, but I wonder what would be the use case for it? For now, setting send buffer capacity to `N*<drain_threshold>` as you suggested before worked perfectly to maintain the concurrency and keeping unit tests fast (when drain threshold is set to 1).

Reply:
You had some code in some place that tried to unblock stuff elsewhere by pulling in something less than a full slice.  Either way, it seems like this might be more general (and this has less state as a result).

Reply:
yea I think that code is not compatible anymore with contiguous buffers. I removed it entirely